logger/logger.swift

```diff
@@ -1,10 +1,188 @@
 //
 //  logger.swift
-//  logger
+//  Test
 //
-//  Created by Daniel Brain on 10/12/20.
+//  Created by Daniel Brain on 10/11/20.
 //
 
-class logger {
+import Foundation
 
+let DEFAULT_PAYLOAD: Payload = [:]
+
+class Logger {
+    private let url: URL
+    private let prefix: String?
+    private let logLevel: LogLevel
+    private let flushInterval: Double
+    
+    private var events = Events()
+    private var headerBuilders: [() -> Headers] = []
+    private var payloadBuilders: [() -> Payload] = []
+    private var trackingBuilders: [() -> Payload] = []
+    
+    init(
+        url : String,
+        prefix : String? = nil,
+        logLevel : LogLevel = DEFAULT_LOGLEVEL,
+        flushInterval: Double = DEFAULT_FLUSH_INTERVAL
+    ) {
+        if (flushInterval <= (FLUSH_DEBOUNCE_TIME * 2)) {
+            preconditionFailure("flushInterval must be greater than \( (FLUSH_DEBOUNCE_TIME * 2) )")
+        }
+        
+        self.url = URL(string: url)!
+        self.prefix = prefix
+        self.logLevel = logLevel
+        self.flushInterval = flushInterval
+        
+        setupFlushInterval()
+    }
```

Comment:
This is more controversial, depending on the project's swift conventions, but I would suggest going with a cleaner init convention. I want to highlight a few changes: 
- The first parameter always has an external name such as `withUrl`, which is beneficial for projects that may have multiple init methods that vary slightly.
- When defining a type, the colon doesn't need a leading and trailing space. `url: String` is slightly cleaner and serves the same purpose as `url : String` with less wasted space overall.
- Typically, try to refrain from line splits unless it is necessary (reduces file length)

```suggestion
    init(withUrl url: String, prefix: String? = nil, logLevel: LogLevel = Config.DEFAULT_LOGLEVEL, flushInterval: Double = Config.DEFAULT_FLUSH_INTERVAL) {
        if (flushInterval <= (Config.FLUSH_DEBOUNCE_TIME * 2)) {
            preconditionFailure("flushInterval must be greater than \( (Config.FLUSH_DEBOUNCE_TIME * 2) )")
        }

        self.url = URL(string: url)!
        self.prefix = prefix
        self.logLevel = logLevel
        self.flushInterval = flushInterval

        setupFlushInterval()
    }
```

Reply:
Good points! Init has the potential to get really long here though, without line splits, right?

Reply:
It really depends on how the init is structured but you can keep an `init` smaller if you pass off less frequent options to a convenience initializer.